ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureBlobDataAccessor.java

```diff
@@ -0,0 +1,306 @@
+        if (!textValue.equals(metadata.get(fieldName))) {
+          metadata.put(fieldName, textValue);
+          // Set condition to ensure we don't clobber another update
+          BlobRequestConditions blobRequestConditions = new BlobRequestConditions().setIfMatch(etag);
+          blobClient.setMetadataWithResponse(metadata, blobRequestConditions, null, Context.NONE);
+        }
+        return true;
+      } finally {
+        storageTimer.stop();
+      }
+    } catch (BlobStorageException e) {
+      if (e.getErrorCode() == BlobErrorCode.CONDITION_NOT_MET) {
+        // TODO: blob was updated (race condition), retry the update
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Permanently delete the specified blobs in Azure storage.
+   * @param blobMetadataList the list of {@link CloudBlobMetadata} referencing the blobs to purge.
+   * @return the number of blobs successfully purged.
+   * @throws BlobStorageException if the purge operation fails.
+   */
+  public int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws BlobStorageException {
+    // TODO: use batch api to delete all
+    // https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/storage/azure-storage-blob-batch
+    throw new UnsupportedOperationException("Not yet implemented");
+  }
+
+  /**
+   * Get the azure blob reference for blobid.
+   * @param blobId id of the blob for which {@code CloudBlockBlob} reference is asked for.
+   * @param autoCreateContainer flag indicating whether to create the container if it does not exist.
+   * @return {@code BlockBlobClient} reference.
+   */
+  private BlockBlobClient getAzureBlobReference(BlobId blobId, boolean autoCreateContainer) {
+    BlobContainerClient containerClient = getContainer(blobId, autoCreateContainer);
+    String azureBlobName = getAzureBlobName(blobId);
+    return containerClient.getBlobClient(azureBlobName).getBlockBlobClient();
+  }
+
+  /**
+   * Get an Azure container to place the specified {@link BlobId}.
+   * @param blobId the {@link BlobId} that needs a container.
+   * @param autoCreate flag indicating whether to create the container if it does not exist.
+   * @return the created {@link BlobContainerClient}.
+   */
+  private BlobContainerClient getContainer(BlobId blobId, boolean autoCreate) {
+    String containerName = getAzureContainerName(blobId.getPartition().toPathString());
+    BlobContainerClient containerClient = storageClient.getBlobContainerClient(containerName);
+    if (autoCreate) {
+      if (!knownContainers.contains(containerName)) {
+        try {
+          if (!containerClient.exists()) {
+            containerClient.create();
+            logger.info("Created container {}", containerName);
+          }
+        } catch (BlobStorageException ex) {
+          if (ex.getErrorCode() != BlobErrorCode.CONTAINER_ALREADY_EXISTS) {
+            logger.error("Failed to create container {}", containerName);
+            throw ex;
```

Comment:
I wonder if this is really an error. Doesn't it just mean that some other thread/vcr got to creating the container first while this thread was doing the existence check?

Reply:
only throws exception if the error is something other than already-exists.